docs/architecture/adr-017-subspace-fee-grant.md

```diff
@@ -0,0 +1,194 @@
+# ADR 017: Subspace fee grant
+
+## Changelog
+
+- Nov 14th, 2022: First draft
+
+## Status
+
+PROPOSED
+
+## Abstract
+
+This ADR introduces a new subspace-based fee grant method, which allows subspace service providers to pay the fees of subspace related transactions for the users.
+
+## Context
+
+Currently, one of the major problems of current Web3 services is complicated to use, it requires users to have tokens paying transaction fees so that they can use the service. For instance, in order to create a post on Desmos-based social network, users have to understand:
+- what `transaction` is and why they need some DSM to broadcast it;
+- get some DSM either via an on-ramp or by swapping existing funds.
+
+The `x/feegrant` gives the possibility to pay fees for the users, meaning that users can use the service without understanding how Web3 works behind. However, the `x/feegrant` allowance is not subspace-specified so subspace fees providers might unexpected pay the fees of the transactions inside another subspace.
+
+## Decision
+
+We will implement a subspace-specified fee grant process base on `x/feegrant` that allows subspace fees providers to pay fees for the users inside the specified subspace. The process of that a subspace fees provider support a user fees will be as follows:
+1. the user asks fees provider a fee grant inside the specified subspace;
+2. the fees provider send a subspace fee grant transaction to Desmos;
+3. Desmos executes the subspace fee grant transaction successfully;
+4. the user can use the service without any tokens inside the specified subspace, the fees will be paid by the fees provider.
+
+```mermaid
+sequenceDiagram
+    participant User
+    participant Fees Provider
+    participant Desmos
+    
+    User->>Fees Provider: 1. ask for a subspace fee grant
+    Fees Provider->>Desmos: 2. send subspace fee grant tx
+    Desmos->>Desmos: 3. execute subspace fee grant tx
+    Desmos-->>Fees Provider:  return ok
+    Fees Provider-->>User: return ok
+
+    User->>Desmos: 4. use app with subspace tx
+    Desmos-->>User: return ok
+```
+
+### DeductFeeDecorator
+
+Currently, `x/auth` provides a `DeductFeeDecorator` based on `x/feegrant` to execute the action deducting fees from the signer/feepayer of a transaction. We will build a new subspace-specified `DeductFeeDecorator` to replace the current one.
+
+The new subspace-specified `DeductFeeDecorator` will operate the fees with the process as follows:
+1. check all the messages in the transaction are the subspace messages and all of them are to the same subspace;
```

Comment:
With this method, could a subspace owner issue a grant for both `MsgSaveProfile` and `MsgCreatePost` messages? If so, could then a user perform a transaction containing one instance of both, without any issue? 

From what I read it seems that these kind of grants are related only to subspace messages, but I think it would be better to allow subspace owners to also issue grants related to `MsgSaveProfile` or other kind of messages since those might be useful when creating a Desmos Profile for the first time within an application.

Reply:
`MsgCreatePost` is allowed, but `MsgSaveProfile` is not since `x/profiles` should be the root module of all desmos modules, additionally, `x/profile` messages are not subspace-specified. We should exclude the `x/profiles` messages from this ADR.

In order to pay the fees of other non-subspace messages like `MsgSaveProfile` for the user in this scenario, the subspace owner can use `x/feegrant` to grant a filtered allowance including `MsgSaveProfile` with spend and periodic limit.

Reply:
I realize that instead of checking if all the messages are the same subspace messages or not, we can specified the subspace id and any needed info in the `memo` field. A format like `subspace-1-0` to use the subspace 1 group 0(default) to check the fee payer.

Reply:
> I realize that instead of checking if all the messages are the same subspace messages or not, we can specified the subspace id and any needed info inside the `memo` field. It would be like a format `subspace-1` to check the fee payer and grant on subspace 1. Then, we can grant any other non-subspace messages. What do you think?

I don't think that's a viable and scalable solution. It would cause the memo field to become unusable for other purposes, limiting the DApps and requiring them to know more things before broadcasting the messages. I think it's better to just check the transaction messages instead

Reply:
Another idea is to wrap the sdk.Tx to desmos.Tx containing the `subspace_id` and `group_id` soo that we can use these field to get the subspace grant easily in the deducting transaction phase. But it will require a new transaction type to increase the maintaining cost.